fms_extras/models/paged_llama.py

```diff
@@ -0,0 +1,811 @@
+models.register_model(_architecture_name, "13b", _llama_factory_factory(_13b_config))
+models.register_model(_architecture_name, "70b", _llama_factory_factory(_70b_config))
+
+
+def _rename_weights_to_fms(orig_sd):
+    replacements = [
+        (r"^tok_embeddings", "headless_model.shared.emb"),
+        (r"^norm", "headless_model.dec_norm"),
+        (r"^output", "head.head"),
+        (r"^layers", "headless_model.layers"),
+        (r"\.attention\.", ".attn."),
+        (r"attn\.wq", "attn.query"),
+        (r"attn\.wk", "attn.key"),
+        (r"attn\.wv", "attn.value"),
+        (r"attn\.wo", "attn.dense"),
+        (r"attention_norm", "ln"),
+        (r"feed_forward\.w1", "ff_sub_layer.wg"),
+        (r"feed_forward\.w2", "ff_sub_layer.w2"),
+        (r"feed_forward\.w3", "ff_sub_layer.w1"),
+        (r"ffn_norm", "ff_ln"),
+    ]
+    new_sd = {}
+    for name, param in orig_sd.items():
+        new_name = name
+        for pattern, repl in replacements:
+            new_name = re.sub(pattern, repl, new_name)
+        new_sd[new_name] = param
+
+    return new_sd
+
+
+def _hf_sd_to_fms_sd(hf_sd: Mapping) -> Mapping:
+    replacements = [
+        (r"^lm_head.weight", "head.head.weight"),
+        (r"^model.embed_tokens.weight", "headless_model.shared.emb.weight"),
+        (r"^model.norm", "headless_model.dec_norm"),
+        (r"^model.layers", "headless_model.layers"),
+        (r"self_attn\.k_proj", "attn.key"),
+        (r"self_attn\.v_proj", "attn.value"),
+        (r"self_attn\.q_proj", "attn.query"),
+        (r"self_attn\.o_proj", "attn.dense"),
+        (r"mlp\.gate_proj", "ff_sub_layer.wg"),
+        (r"mlp\.up_proj", "ff_sub_layer.w1"),
+        (r"mlp\.down_proj", "ff_sub_layer.w2"),
+        (r"input_layernorm", "ln"),
+        (r"post_attention_layernorm", "ff_ln"),
+    ]
+    new_sd = {}
+
+    trans_required_pattern = re.compile(
+        "headless_model.layers.[0-9]+.attn.(query|key).weight"
+    )
+    for name, param in hf_sd.items():
+        new_name = name
+        for pattern, repl in replacements:
+            new_name = re.sub(pattern, repl, new_name)
+        new_sd[new_name] = param
+
+        # hf -> fms requires a transpose operation for the query and key
+        if bool(trans_required_pattern.match(new_name)):
+            temp = new_sd[new_name]
+            # nheads is used in the transformation required for hf->fms
+            # here we are using 128 as this value fits with all popular models
+            #   7B, 13B, 70B to recover the number of heads
+            nheads = int(temp.size(0) / 128)
+
+            temp = (
+                temp.view(nheads, 2, -1, temp.size(1))
+                .transpose(1, 2)
+                .reshape(*temp.size())
+            )
+
+            new_sd[new_name] = temp
+
+    return new_sd
+
+
+serialization.register_adapter(_architecture_name, "meta", _rename_weights_to_fms)
+serialization.register_adapter(_architecture_name, "hf", _hf_sd_to_fms_sd)
```

Comment:
if the adapters are the same maybe we can just register_adapter(name, source, get_adapter(original_name, source)) and avoid the extra (copy/paste?) adapter functions

Reply:
They are slightly different because there is now a headless model for PagedLLaMA so the mapping required an extra `headless_model` prefix. The logic is basically all the same though.

Reply:
Aside from this, I am wondering if it would be valuable to add a `SimpleMapping` adapter that will take in a list of tuples - replacements. This way, rather than re-implementing an adapter, we just do something like:

```python
meta_replacements = [
        (r"^lm_head.weight", "shared.head.weight"),
        ...
        (r"post_attention_layernorm", "ff_ln"),
    ]

serialization.register_adapter("llama", "meta", SimpleMapping(meta_replacements))
```

It saves a small amount of duplication, and may be simpler for people to do

Reply:
that might be nice if it works, though we do have some transposes and such here too so not sure when it would be useful.

I'm not loving having so much code copy/pasted from the main repo with small tweaks but I guess this is supposed to be temporary until we implement PagedTensor?

Another difference is that we're using fused weights here, which we haven't landed in the main repo yet either.